app/models/events/tags.rb

```diff
@@ -0,0 +1,39 @@
+module Events::Tags
+  extend ActiveSupport::Concern
+
+  included do
+    validates :tag, presence: true, inclusion: { in: ->(_) { tags.keys } }
+    attribute :tag, Events::Tag::Type.new
+    validates :additional, nested: true, allow_nil: true
+    belongs_to :ip_address, optional: true
+    belongs_to :geoip_info, optional: true
+
+    cattr_reader(:tags) { {} }
+
+    def additional_type
+      tags.fetch(tag, nil)
+    end
+
+    def additional
+      additional_type&.new(super) || super
+    end
+
+    def additional=(value)
+      super(value&.to_h)
+    end
```

Comment:
I suggest adding these methods to the module rather than in `included do`. I think it's better for the method lookup table, so they can be shared by the module instead of unique per singleton.